File: PBMovie/src/main/java/project/threads/SendThread.java
package project.threads;

/**
 * SendThread: Thread para envio de requisições ao TCPServer 
 */

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.IOException;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import project.stream.Parser;
import project.javaOut.Movie;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
import project.stream.Format;

public class SendThread extends Thread {
    DataInputStream input;

    DataOutputStream output;

    Socket socket;
    
    //Logger logger = LoggerFactory.getLogger(SendThread.class);

    public SendThread(Socket socket) {
        try {
            this.socket = socket;
            this.input = new DataInputStream(socket.getInputStream());
            this.output = new DataOutputStream(socket.getOutputStream());
        } catch (IOException ioe) {
        	//logger.warn("IOException: " + ioe.getMessage());
            System.out.println("IOException: " + ioe.getMessage());
        } //catch
        
    } //construtor
    
    void handleCodeOperation(String operation, Scanner scanner) throws IOException { 
        Format format = new Format();
        Parser parser = new Parser();

        switch (operation) {
            case "create":
                // create 
                Movie movie = formMovie(scanner);
                byte[] createRequest = format.formatCreate(movie);
                this.output.write(createRequest, 0, createRequest.length);  // envia a mensagem para o servidor
                break;
            case "read":
                System.out.print("Nome do filme: ");
                String movieName = scanner.nextLine();
                byte[] readRequest = format.formatRead(movieName);
                this.output.write(readRequest, 0, readRequest.length);
                break;
            case "update":
                // Mandar read, pegar Movie e fazer um form com os valores correntes do lado
                // os que forem preenchidos serao mandados na requisição, sem ser um objeto 
                // protobuff.
                break;
            case "delete":
                System.out.print("Nome do filme: ");
                String movieNameDel = scanner.nextLine();
                byte[] deleteRequest = format.formatDelete(movieNameDel);
                this.output.write(deleteRequest, 0, deleteRequest.length);
                break;
            case "listByActor":
                System.out.print("Nome do ator: ");
                String actorName = scanner.nextLine();
                byte[] listByActorRequest = format.formatListByActor(actorName);
                this.output.write(listByActorRequest, 0, listByActorRequest.length);
                byte[] listByActorResponse = new byte[1024]; 
                this.input.read(listByActorResponse);
            default:
                System.out.println("ERROR: Unsupported operation\n");
                break;
        }
    } //connectUser

    Movie formMovie(Scanner scanner) {
        Movie.Builder movie = Movie.newBuilder();

        System.out.println("\nTitle: ");
        String title = scanner.nextLine(); 
        movie.setTitle(title);

        System.out.println("\nYear: ");
        int year = Integer.valueOf(scanner.nextLine()); 
        movie.setYear(year);

        System.out.println("\nDate: ");
        String released = scanner.nextLine(); 
        movie.setReleased(released);

        System.out.println("\nURL Poster: ");
        String poster = scanner.nextLine(); 
        movie.setPoster(poster);

        System.out.println("\nPlot: ");
        String plot = scanner.nextLine(); 
        movie.setPlot(plot);           

        System.out.println("\nFull plot: ");
        String fullplot = scanner.nextLine(); 
        movie.setFullplot(fullplot);           

        System.out.println("Directors' names:\n");
        System.out.println("(Press key 'q' for quit)\n");
        
        int countDirectors = 1;
        while(true) {
            System.out.format("\nDirector %d: ", countDirectors);
            String directorName = scanner.nextLine(); 

            if ("q".equals(directorName)) {
                if (countDirectors == 1) {
                   System.out.println("ERROR: At least one director must be provided\n"); 
                   continue;
                } else {
                    System.out.println("Quitting...\n");
                    break;
                }
            }

            movie.addDirectors(directorName);
            countDirectors++;
        } 

        System.out.println("Actors' names:\n");
        System.out.println("(Press key 'q' for quit)\n");

        int countCast = 1;
        while(true) {
            System.out.format("\nActor %d: ", countCast);
            String actorName = scanner.nextLine(); 

            if ("q".equals(actorName)) {
                if (countCast == 1) {
                   System.out.println("ERROR: At least one actor must be provided\n"); 
                   continue;
                } else {
                    System.out.println("Quitting...\n");
                    break;
                }
            }

            movie.addCast(actorName);
            countCast++;
        }

        System.out.println("Countries' names:\n");
        System.out.println("(Press key 'q' for quit)\n");

        int countCountry = 1;
        while(true) {
            System.out.format("\nCountry %d: ", countCountry);
            String countryName = scanner.nextLine(); 

            if ("q".equals(countryName)) {
                if (countCountry == 1) {
                   System.out.println("ERROR: At least one country must be provided\n"); 
                   continue;
                } else {
                    System.out.println("Quitting...\n");
                    break;
                }
            }

            movie.addContries(countryName);
            countCountry++;
        }

        System.out.println("Genres' names:\n");
        System.out.println("(Press key 'q' for quit)\n");

        int countGenre = 1;
        while(true) {
            System.out.format("\nGenre %d: ", countGenre);
            String genreName = scanner.nextLine(); 

            if ("q".equals(genreName)) {
                if (countGenre == 1) {
                   System.out.println("ERROR: At least one genre must be provided\n"); 
                   continue;
                } else {
                    System.out.println("Quitting...\n");
                    break;
                }
            }

            movie.addGenres(genreName);
            countGenre++;
        }

        return movie.build();
    }

    /* metodo executado ao iniciar a thread - start() */
    @Override
    public void run() {
        Scanner scanner = new Scanner(System.in); // ler mensagens via teclado

        try {
            String buffer = "";
            while (true) {
                buffer = scanner.nextLine(); // lê mensagem via teclado
                
                if (buffer.trim().equals("EXIT")) {
                	//logger.info("Client exiting");
                	break;
                }
                
                handleCodeOperation(buffer, scanner);
            }

            System.out.println("SendThread exitted\n");
        } catch (EOFException eofe) {
        	//logger.warn("EOFException: " + eofe.getMessage());
            System.out.println("EOFException: " + eofe.getMessage());
        } catch (IOException ioe) {
        	//logger.warn("IOException: " + ioe.getMessage());
            System.out.println("IOException: " + ioe.getMessage());
        } finally {
            try {
                this.input.close();
                this.output.close();
                this.socket.close();
            } catch (IOException ioe) {
            	//logger.warn("IOException: " + ioe.getMessage());
                System.err.println("IOException: " + ioe);
            }
        }
        //logger.info("SendThread finished.");
        System.out.println("SendThread finished.");
    } //run
} //class

File: PBMovie/src/main/kotlin/project/threads/SendThread.kt
package project.threads

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.EOFException
import java.io.IOException
import java.net.Socket
import java.util.ArrayList
import kotlin.collections.List
import java.util.Scanner
import org.bson.json.JsonObject
import java.io.StringWriter
import org.bson.json.JsonWriter
import project.javaOut.Movie
import project.stream.FormatRequest
//import org.slf4j.Logger
//import org.slf4j.LoggerFactory

/**
 * SendThread: Thread para envio de requisições ao ReceiveThreadServer 
 */
public class SendThread(socketC: Socket): Thread() {
    val input: DataInputStream
    val output: DataOutputStream
    val formatReq: FormatRequest
    val socket: Socket

    init {
        formatReq = FormatRequest()
        socket = socketC
        input = DataInputStream(socket.getInputStream())
        output = DataOutputStream(socket.getOutputStream())
    }

    /**
     * Faz procedimentos necessários para formatação e envio da requisição com formulários e FormatRequest
     * @param operation String correspondente a operação desejada
     * @param scanner Scanner para leitura de entradas por terminal
     * @return Envio da requisição
     */
    @Throws(IOException::class)
    fun handleCodeOperation(operation: String, scanner: Scanner) { 
        when (operation) {
            "create" -> {
                var movie: Movie = formMovie(scanner)
                var createRequest: ByteArray = formatReq.create(movie)
                output.write(createRequest, 0, createRequest.size)
            }
            "read" -> {
                System.out.print("Movie name: ")
                var movieName: String = scanner.nextLine()
                var readRequest: ByteArray = formatReq.read(movieName)
                output.write(readRequest, 0, readRequest.size)
            }
            "update" -> {
                var movieUpdates: Movie = formEditMovie(scanner)

                var updateRequest: ByteArray = formatReq.update(movieUpdates)
                output.write(updateRequest, 0, updateRequest.size)
            }
            "delete" -> {
                System.out.print("Movie name: ")
                var movieNameDel: String = scanner.nextLine()
                var deleteRequest: ByteArray = formatReq.delete(movieNameDel)
                output.write(deleteRequest, 0, deleteRequest.size)
            }
            "listByActor" -> {
                System.out.print("Actor name: ")
                var actorName: String = scanner.nextLine()
                var listByActorRequest: ByteArray = formatReq.listByAttribute(5, actorName)
                output.write(listByActorRequest, 0, listByActorRequest.size)
            }
            "listByGenre" -> {
                System.out.print("Genre name: ")
                var genreName: String = scanner.nextLine()
                var listByGenreRequest: ByteArray = formatReq.listByAttribute(6, genreName)
                output.write(listByGenreRequest, 0, listByGenreRequest.size)
            }
            else -> {
                System.out.println("ERROR: Unsupported operation\n")
            }
        }
    }

    /**
     * Formulário para criação de um Movie. Não são aceitos campos vazios
     * @param scanner Scanner para receber entradas do terminal
     * @return Objeto Movie com os valores fornecidos 
     */
    @Throws(IOException::class)
    fun formMovie(scanner: Scanner): Movie {
        var movie: Movie.Builder = Movie.newBuilder()

        System.out.print("Title: ")
        var title: String = scanner.nextLine() 
        if (title.isBlank()) {
            throw IOException("Title cannot be blank")
        }
        movie.setTitle(title)

        System.out.print("\nYear: ")
        var year: Int = Integer.valueOf(scanner.nextLine()) 
        movie.setYear(year)

        System.out.print("\nDate: ")
        var released: String = scanner.nextLine() 
        movie.setReleased(released)

        System.out.print("\nURL Poster: ")
        var poster: String = scanner.nextLine() 
        movie.setPoster(poster)

        System.out.print("\nPlot: ")
        var plot: String = scanner.nextLine() 
        movie.setPlot(plot)           

        System.out.print("\nFull plot: ")
        var fullplot: String = scanner.nextLine() 
        movie.setFullplot(fullplot)           

        System.out.println("\nDirectors' names:\n")
        System.out.println("(Press key 'q' for quit)\n")
        
        var countDirectors = 1
        while(true) {
            System.out.format("\nDirector %d: ", countDirectors)
            var directorName: String = scanner.nextLine() 

            if ("q".equals(directorName)) {
                if (countDirectors == 1) {
                   System.out.println("ERROR: At least one director must be provided\n") 
                   continue
                } else {
                    break
                }
            }

            movie.addDirectors(directorName)
            countDirectors++
        } 

        System.out.println("\nActors' names:\n")
        System.out.println("(Press key 'q' for quit)\n")

        var countCast = 1
        while(true) {
            System.out.format("\nActor %d: ", countCast)
            var actorName: String = scanner.nextLine() 

            if ("q".equals(actorName)) {
                if (countCast == 1) {
                   System.out.println("ERROR: At least one actor must be provided\n") 
                   continue
                } else {
                    break
                }
            }

            movie.addCast(actorName)
            countCast++
        }

        System.out.println("\nCountries' names:\n")
        System.out.println("(Press key 'q' for quit)\n")

        var countCountry = 1
        while(true) {
            System.out.format("\nCountry %d: ", countCountry)
            var countryName: String = scanner.nextLine() 

            if ("q".equals(countryName)) {
                if (countCountry == 1) {
                   System.out.println("ERROR: At least one country must be provided\n") 
                   continue
                } else {
                    break
                }
            }

            movie.addContries(countryName)
            countCountry++
        }

        System.out.println("\nGenres' names:\n")
        System.out.println("(Press key 'q' for quit)\n")

        var countGenre = 1
        while(true) {
            System.out.format("\nGenre %d: ", countGenre)
            var genreName: String = scanner.nextLine() 

            if ("q".equals(genreName)) {
                if (countGenre == 1) {
                   System.out.println("ERROR: At least one genre must be provided\n") 
                   continue
                } else {
                    break
                }
            }

            movie.addGenres(genreName)
            countGenre++
        }

        return movie.build()
    }

    fun formEditMovie(scanner: Scanner): Movie {
        var movie: Movie.Builder = Movie.newBuilder()

        System.out.print("Title: ")
        var title: String = scanner.nextLine() 
        movie.setTitle(title)

        System.out.print("\nYear: ")
        try {
            var yearString: String = scanner.nextLine()
            if (yearString.isNotBlank()) {
                var year: Int = Integer.valueOf(yearString) 
                movie.setYear(year)
            }
        } catch (nfe: NumberFormatException) {
            throw IOException("Expected a number")
        }

        System.out.print("\nDate: ")
        var released: String = scanner.nextLine() 
        if (released.isNotBlank()) {
            movie.setReleased(released)
        }

        System.out.print("\nURL Poster: ")
        var poster: String = scanner.nextLine() 
        if (poster.isNotBlank()) {
            movie.setPoster(poster)
        }

        System.out.print("\nPlot: ")
        var plot: String = scanner.nextLine() 
        if (plot.isNotBlank()) {
            movie.setPlot(plot)           
        }

        System.out.print("\nFull plot: ")
        var fullplot: String = scanner.nextLine() 
        if (fullplot.isNotBlank()) {
            movie.setFullplot(fullplot)           
        }

        System.out.println("\nDirectors' names:\n")
        System.out.println("(Press key 'q' for quit)\n")
        
        var countDirectors = 1
        while(true) {
            System.out.format("\nDirector %d: ", countDirectors)
            var directorName: String = scanner.nextLine() 

            if ("q".equals(directorName) || directorName.isBlank()) {
                break
            }

            movie.addDirectors(directorName)
            countDirectors++
        } 

        System.out.println("\nActors' names:\n")
        System.out.println("(Press key 'q' for quit)\n")

        var countCast = 1
        while(true) {
            System.out.format("\nActor %d: ", countCast)
            var actorName: String = scanner.nextLine() 

            if ("q".equals(actorName) || actorName.isBlank()) {
                break
            }

            movie.addCast(actorName)
            countCast++
        }

        System.out.println("\nCountries' names:\n")
        System.out.println("(Press key 'q' for quit)\n")

        var countCountry = 1
        while(true) {
            System.out.format("\nCountry %d: ", countCountry)
            var countryName: String = scanner.nextLine() 

            if ("q".equals(countryName) || countryName.isBlank()) {
                break
            }

            movie.addContries(countryName)
            countCountry++
        }

        System.out.println("\nGenres' names:\n")
        System.out.println("(Press key 'q' for quit)\n")

        var countGenre = 1
        while(true) {
            System.out.format("\nGenre %d: ", countGenre)
            var genreName: String = scanner.nextLine() 

            if ("q".equals(genreName) || genreName.isBlank()) {
                break
            }

            movie.addGenres(genreName)
            countGenre++
        }

        return movie.build()
    }

    public override fun run() {
        var scanner = Scanner(System.`in`)

        try {
            var buffer: String
            while (true) {
                buffer = scanner.nextLine()
                print("\n")

                if (buffer.trim().equals("EXIT")) {
                    //logger.info("Client exiting")
                    break
                }

                try {
                    handleCodeOperation(buffer, scanner)
                } catch (ioe: IOException) {
                    println("\nERROR: " + ioe.message + "\n")
                } catch (e: Exception) {
                    println("\nERROR: " + e.message + "\n")
                }
            }
        } catch (eofe: EOFException) {
            //logger.warn("EOFException: " + eofe.message)
            System.out.println("EOFException: " + eofe.message)
        } catch (ioe: IOException) {
            //logger.warn("IOException: " + ioe.message)
            System.out.println("IOException: " + ioe.message)
        } finally {
            try {
                input.close()
                output.close()
                socket.close()
            } catch (ioe: IOException) {
                //logger.warn("IOException: " + ioe.message)
                System.err.println("IOException: " + ioe.message)
            }
        }
        //logger.info("SendThread finished.")
        System.out.println("SendThread finished.")
    }
}

